### DISCUSSION TITLE
Proxy Yield Aggregator v1.0.0 - ERC4626 Compatible Deposit Contract

### DISCUSSION THREAD

**Discussion by 0x4007** [pavlovcik/uad-ui#171] [Protocol Specifications]
# "Bring Your Own Yield Source"

For the next iteration of the Proxy Yield Aggregator (PYA) program a user should be able to bring their own [ERC4626](https://github.com/fei-protocol/ERC4626/blob/7a947f2507b760ae470578cfb106f71ff5b1a14b/src/ERC4626Router.sol) compatible yield source, and connect it to our deposit contract. 

This product is intended for users to be able to stack the total yield that they can receive from their favorite yield bearing sources. 

Upon withdrawal, the user interface should present a slider which represents risk tolerance of the user. 

## Example

1. The user likes a new ERC4626 Compatible Yearn USDC vault that promises 20% APR.
1. The user "connects" the ERC4626 vault to our deposit contract. 
    1. Implementation details must be researched, but I hope that we can avoid a contract redeployment, a la deploying a new Uniswap market.
1. The user deposits 1000 USDC, and the USDC is immediately transferred to the underlying yield source.
1. One year later, the user seeks to withdraw their 1200 USDC, and returns to our user interface to withdraw.
1. The user interface presents a slider, which seeks to understand the user's risk/reward tolerance 
    1. For example, the user can withdraw with no additional risk (1200 USDC) or the user can withdraw with max risk (2400 uCR)
    1. The idea is that the user can convert the USDC into our uCR for a premium 
    1. In exchange, the Ubiquity Dollar is able to raise collateral for the dollar in perpetuity.
1. User withdraws at "25% r/r" netting them something like 900 USDC and 900 uCR (premium curve formula to be determined)
    1. "50% r/r" would be 600 USDC and 1200 uCR
    1. "100% r/r" is 0 USDC and 2400 uCR
    1. "16.7% r/r" is 1000 USDC and 400 uCR // I imagine that many would, in this example go here or lower in order to only convert yield, and to receive the principle back.

## Adjustable Parameters

1. uCR Premium, a math formula (default would be a linear calculation, 1 USDC -> 2 uCR)
    1. This would be deployed as a separate smart contract
    1. This gives us flexibility to implement exponential or logarithmic functions which may more efficiently produce our desired economic outcomes!
1. Yield token whitelist (default would be an array with the following contract addresses: USDC, DAI, LUSD)
    1. The idea is that, because the user is able to bring their own yield source, we wouldn't want a situation where they give assets that we do not want (like UST)
    1. Rejection should happen upon deposit, not withdrawal; in case we change our accepted assets after the user already has deposited their tokens.

## Remarks

It would be very interesting to match the uCR Premium (math formula) with the bond of the associated asset. That way we are purchasing at the same price premiums ecosystem wide. As long as we're mindful of 1. the interface implementation, and 2. that we architect it so we can associate the uCR Premium PER asset, then I think we can easily share the same formula between our bonds and the PYA.

**Comment by sergfeldman** [comment]
The image from Seo

![image_2022-05-25_02-55-40](https://user-images.githubusercontent.com/15567379/170111159-0b350efc-9a76-46d3-916c-5b1de6e818d4.png)

**Comment by 0x4007** [comment]
Seems handy

https://twitter.com/jameskbh/status/1528862880988536833?s=21&t=q0eTq5voI87HiF_7nRxLVw

**Comment by zapaz** [comment]
> 1 Is it necessary to have one vault for each of the three assets or is it necessary to deposit three assets in one vault, which causes problems when depositing and withdrawing.
> 
> Without researching I assume we would need to have users mint new vaults like Uniswap markets. I also assume that each vault should support only one asset at a time, as each vault is associated with only one underlying yield bearing source/strategy.

Good solution is to use minimal proxy [EIP-1167](https://eips.ethereum.org/EIPS/eip-1167) 
With Open Zeppelin library [EIP-1167 lib](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/proxy/Clones.sol)
And one example using this library, with template capabilities added, I made earlier : [CloneFactory](https://github.com/Kredeum/kredeum/blob/beta/hardhat/contracts/CloneFactoryV2.sol)

**Comment by zapaz** [comment]
```
Smart contract overview.

1. Vault contract : only one assets can be deal in this contract. 
	a. Can save interest-bearing assets or forward to users 
	b. Able to connect with several Strategies brought by users
	c. ERC4626 standard and compatible
	d. Based ERC4626 mint uCR and corresponding assets flows into Admin Trusted Wallet
	e. uCR token contract	
2. Vault setting contract
	a. Set Assets whitelist
	b. Interfaces of each Assets
	c. Functions for Admin; Pause, Unpause, Treasury wallet setting, 
3. Oracle and Proxy

1a: I don't understand why you're highlighting this. I have not seen yield aggregators work with interest-bearing assets e.g. AAVE USDC and Compound USDC. I have only seen them with normal assets e.g. USDC.
1e. [uAR](https://etherscan.io/address/0x5894cFEbFdEdBe61d01F20140f41c5c49AedAe97) was recently rebranded to uCR
3. There is not enough detail here for me to understand.

```

same remarks than @pavlovcik  : 
 
**1a and 1b** : seems the purpose of [ERC4626](https://eips.ethereum.org/EIPS/eip-4626) is to hide this complexity, and it seems not part of this smartcontract

**1c and 1d** : 👍🏻   

**3** : can you detail the Oracle part ?

**Comment by 0x4007** [comment]
# Review Checklist 

Let's go through these and explain to me how the [current implementation](https://github.com/ubiquity/ubiquity-dollar/blob/feat/foundry-conversion/packages/contracts/src/dollar/ProxyYieldAggregator.sol) handles these. Ideally there are test cases that correlate with each of these, but unfortunately that does not appear to be the case.

```log
➤ YN0000: Running 7 tests for test/dollar/ProxyYieldAggregator.t.sol:ProxyYieldAggregatorTest
➤ YN0000: [PASS] testDepositWithStrategyAsset() (gas: 188416)
➤ YN0000: [PASS] testDepositWithStrategyAsset_And_Redeem() (gas: 308634)
➤ YN0000: [PASS] testDepositWithStrategyAsset_And_RedeemStrategyAsset() (gas: 319129)
➤ YN0000: [PASS] testDepositWithStrategyAsset_And_RedeemWithSplit() (gas: 308609)
➤ YN0000: [PASS] testDepositWithStrategyAsset_And_Withdraw() (gas: 309726)
➤ YN0000: [PASS] testDepositWithStrategyAsset_And_WithdrawStrategyAsset() (gas: 320398)
➤ YN0000: [PASS] testDepositWithStrategyAsset_And_WithdrawWithSplit() (gas: 309903)
➤ YN0000: Test result: ok. 7 passed; 0 failed; finished in 23.36s
```

@0xcodercrane @zgorizzo69 

- [x] The user likes a new ERC4626 Compatible Yearn USDC vault that promises 20% APR.
- [x] The user "connects" the ERC4626 vault to our deposit contract. 
    - [x] Implementation details must be researched, but I hope that we can avoid a contract redeployment, a la deploying a new Uniswap market.
- [x] The user deposits 1000 USDC, and the USDC is immediately transferred to the underlying yield source.
- [x] One year later, the user seeks to withdraw their 1200 USDC, and returns to our user interface to withdraw.
- [x] The user interface presents a slider, which seeks to understand the user's risk/reward tolerance 
    - [x] For example, the user can withdraw with no additional risk (1200 USDC) or the user can withdraw with max risk (2400 uCR)
    - [x] The idea is that the user can convert the USDC into our uCR for a premium 
    - [x] In exchange, the Ubiquity Dollar is able to raise collateral for the dollar in perpetuity.
- [x] User withdraws at "25% r/r" netting them something like 900 USDC and 900 uCR (premium curve formula to be determined)
    - [x] "50% r/r" would be 600 USDC and 1200 uCR
    - [x] "100% r/r" is 0 USDC and 2400 uCR
    - [x] "16.7% r/r" is 1000 USDC and 400 uCR // I imagine that many would, in this example go here or lower in order to only convert yield, and to receive the principle back.

## Adjustable Parameters

- [ ] uCR Premium, a math formula (default would be a linear calculation, 1 USDC -> 2 uCR)
    - [ ] This would be deployed as a separate smart contract
    - [ ] This gives us flexibility to implement exponential or logarithmic functions which may more efficiently produce our desired economic outcomes!
- [ ] Yield token whitelist (default would be an array with the following contract addresses: USDC, DAI, LUSD)
    - [ ] The idea is that, because the user is able to bring their own yield source, we wouldn't want a situation where they give assets that we do not want (like UST)
    - [ ] Rejection should happen upon deposit, not withdrawal; in case we change our accepted assets after the user already has deposited their tokens.

**Comment by 0x4007** [comment]
- Make new manager contract with "minimum split" 
- call function on manager to deploy new proxy yield aggregator
- - add to manager array of pyas then we can query from UI to see all "community vaults"

**Reply by zgorizzo69** [reply]
- on withdrawal what to do with treasury money
  - dumping to eth 
  - arbitrage bot that swap when peg > 1$ 
  - swapping with cowswap
- proxy yield manager 
  - a custom calculation per proxy yield address
  - add/remove token allowed to be deposited aka deposit whitelist
  - one transaction call to deploy a proxy yield (factory) with minimum split + grant the UBQ minter right + enforce an underlying asset whitelist

**Reply by 0x4007** [reply]
We also need help with figuring out what to do upon withdrawal. The issue is that as of the current design: we have a deposit whitelist (assets we trust, for example ` [ USDC, USDT, DAI, FRAX, LUSD, RAI, ETH ] `) however we only want to hold censorship resistant assets to back our dollar ` [ LUSD, RAI, ETH ] ` ideally the more stable, the better (LUSD).

The issue is that when the user withdraws, we need to call a swap. If the swap fails for any reason, the transaction is reverted. 

How do we guarantee that when people use our vaults, that we will automatically end up holding LUSD (ideally?)

**Reply by cuddle-dev** [reply]
> add to manager array of pyas then we can query from UI to see all "community vaults"

I am leaning to use the subgraph to query the vaults on the UI.

**Reply by 0x4007** [reply]
> > add to manager array of pyas then we can query from UI to see all "community vaults"
> 
> I am leaning to use the subgraph to query the vaults on the UI.

@zgorizzo69 expressed a concern about gas but I thought that its free to read data off chain so why bother with the overhead of a subgraph if theres just one method to call, with one array as a return value

**Reply by 0x4007** [reply]
@zgorizzo69 we could use FRAXSWAP TWAMM to do long, slow purchases

https://docs.frax.finance/fraxswap/protocol-uses-and-dao-to-dao-swaps

**Reply by 0x4007** [reply]
I just tried FRAX->FXS 7 day purchase and the execution rate is pretty bad. It's approximately 3% slippage.
<img width="130" alt="Screen Shot 2022-10-13 at 19 56 59" src="https://user-images.githubusercontent.com/4975670/195589930-9d4ac4c9-f49f-4a19-b164-99ecce94f2f8.png">
<img width="788" alt="Screen Shot 2022-10-13 at 19 57 20" src="https://user-images.githubusercontent.com/4975670/195589978-274ad6bd-7572-4edc-9085-0157c1636354.png">

**Reply by 0x4007** [reply]
cowswap twamm? 

https://github.com/nlordell/cow-twap-order
https://ethglobal.com/showcase/twap-til-the-cows-come-home-7igm0

**Reply by zgorizzo69** [reply]
> cowswap twamm?
> 
> https://github.com/nlordell/cow-twap-order https://ethglobal.com/showcase/twap-til-the-cows-come-home-7igm0

that seems to be a really good idea but I need to dig deeper and test

**Reply by zgorizzo69** [reply]
> * on withdrawal what to do with treasury money
>       
>       * dumping to eth
>       * arbitrage bot that swap when peg > 1$
>       * swapping with cowswap
> 
>     * proxy yield manager
>       
>       * a custom calculation per proxy yield address
>       * add/remove token allowed to be deposited aka deposit whitelist
>       * one transaction call to deploy a proxy yield (factory) with minimum split + grant the UBQ minter right + enforce an underlying asset whitelist

@pavlovcik do you want me to work on the proxy manager/factory yet ?

**Reply by 0x4007** [reply]
> > * on withdrawal what to do with treasury money
> 
> >       
> 
> >       * dumping to eth
> 
> >       * arbitrage bot that swap when peg > 1$
> 
> >       * swapping with cowswap
> 
> > 
> 
> >     * proxy yield manager
> 
> >       
> 
> >       * a custom calculation per proxy yield address
> 
> >       * add/remove token allowed to be deposited aka deposit whitelist
> 
> >       * one transaction call to deploy a proxy yield (factory) with minimum split + grant the UBQ minter right + enforce an underlying asset whitelist
> 
> 
> 
> @pavlovcik do you want me to work on the proxy manager/factory yet ?

Yes although I made another issue related to upgrading the UbiquityAlgorithmicDollarManager.sol to the Diamond Standard so perhaps consider building this manager in a way that will be easy for a dev to merge into easily?

**Reply by 0x4007** [reply]
Another cowswap related tech

https://twitter.com/hasufl/status/1594084866370416641?s=46&t=SUuHlxGBxLaLBmhyOiMz9g

https://medium.com/iearn/yearn-cow-swap-371b6d7cf3b3

**Reply by 0x4007** [reply]
I think we should consider TWAMMing into LUSD instead of ETH.

**Reply by zgorizzo69** [reply]
here is a small report based on our latest discussion

# Report swapping treasury yield rewards to $LUSD 

## CowSwap
Cowswap offers a mechanism to let contracts "signed" a transaction. Indeed by triggering `setPreSignature` SC can approve a trade by providing its UID.

![image](https://user-images.githubusercontent.com/16853695/205271160-18db83b1-2b27-49bc-938a-9a89a9d07200.png)

the problem is that we need to make a call to the api to get that UID and that is what the gnosis frontend is doing.

![image](https://user-images.githubusercontent.com/16853695/205271347-f59cebd8-ab5e-44db-9860-217ca53fe95c.png)

It is therefore not feasible synchronously onchain 

## adding reward token to lending pool 

we could provide the reward to lending pool that will generate yield and can be used as a collateral to take a loan  issued in $LUSD

drawbacks:
* loan needs to be overcollateralized
* need to find the right lending pool for different reward tokens
* we are still exposed to the reward token 

## thinking outside the box 

Here we want to avoid MEV in the event of a large swap so basically we want to avoid being sandwiched by an attacker.
In order to be profitable the swap would need to be large ( IMO > 100k) and gas low. Lately the [sandwich activity went down](https://www.tarlogic.com/blog/ethereum-blockchain-sandwich-attacks/) because of these factors and thanks to counter measure like flashbot and MEV proof RPC.

We could have a max amount to be swapped  for the treasury and mesure /analyse to detect if we are being scandwich and if we have leftover to be swapped

**Reply by 0x4007** [reply]
> we could provide the reward to lending pool

I think the risk/reward on lending is not great. There are better opportunities with LP swap fees on UniV3.

I think it could be cool to "make up for the difference" over many swaps after. For example: 
1. a whale dumps $1M
2. We inject an extra $100k on top of every sell transaction that occurs after, until we cover the full deficit.

**Comment by 0x4007** [comment]
These guys are aggregating all the ERC4626 compliant vaults in one place. We should consider building off of it?
https://twitter.com/superformxyz/status/1592527993230225409

**Reply by zgorizzo69** [reply]
we could add our vault to their permissionless vault listing
